src/R2N.jl

```diff
@@ -0,0 +1,387 @@
+      max_eval = max_eval,
+      iter = stats.iter,
+      max_iter = max_iter,
+      max_time = max_time,
+    ),
+  )
+
+  callback(nlp, solver, stats)
+
+  done = stats.status != :unknown
+  cgtol = max(rtol, min(T(0.1), √norm_∇fk, T(0.9) * cgtol))
+  
+  local ρk
+  while !done
+    ∇fk .*= -1
+    subsolve!(solver, s, zero(T), n, subsolver_verbose)
+
+    slope = dot(s, ∇fk) # = -dot(s, ∇fk) but ∇fk is negative
+    mul!(Hs, H, s)
+    curv = dot(s, Hs)
+
+    ΔTk = slope + curv / 2
+    ck .= x .+ s
+    fck = obj(nlp, ck)
+
+    if non_mono_size > 1  #non-monotone behaviour
+      k = mod(stats.iter, non_mono_size) + 1
+      solver.obj_vec[k] = stats.objective
+      fck_max = maximum(solver.obj_vec)
+      ρk = (fck_max - fck) / (fck_max - stats.objective + ΔTk)
+    else
+      ρk = (stats.objective - fck) / ΔTk
+    end
+
+    # Update regularization parameters and Acceptance of the new candidate
+    step_accepted = ρk >= η1 && σk >= η2
+    if step_accepted
+      μk = max(μmin, μk / λ)
+      x .= ck
+      grad!(nlp, x, ∇fk)
+      if isa(nlp, QuasiNewtonModel)
+        ∇fn .-= ∇fk
+        ∇fn .*= -1  # = ∇f(xₖ₊₁) - ∇f(xₖ)
+        push!(nlp, s, ∇fn)
+        ∇fn .= ∇fk
+      end
+      set_objective!(stats, fck)
+      unbounded = fck < fmin
+      norm_∇fk = norm(∇fk)
+    else
+      μk = μk * λ
+    end
+
+    set_iter!(stats, stats.iter + 1)
+    set_time!(stats, time() - start_time)
+
+    cgtol = max(rtol, min(T(0.1), √norm_∇fk, T(0.9) * cgtol))
+    set_dual_residual!(stats, norm_∇fk)
+
+    solver.σ = σk
+    solver.μ = μk
+    solver.cgtol = cgtol
+
+    callback(nlp, solver, stats)
+
+    norm_∇fk = stats.dual_feas # if the user change it, they just change the stats.norm 
+    μk = solver.μ
+    cgtol = solver.cgtol
+    σk = μk * norm_∇fk
+
+    optimal = norm_∇fk ≤ ϵ
+
+    if verbose > 0 && mod(stats.iter, verbose) == 0 #TODO change it in the verbose callback #TODO turn off and on as a flag in the callback
+      @info infoline
+      σ_stat = step_accepted ? "↘" : "↗"
+      infoline =
+        @sprintf "%5d  %9.2e  %7.1e  %7.1e  %7.1e  %+7.1e  %1s" stats.iter stats.objective norm_∇fk μk σk ρk σ_stat
+    end
+    if stats.status == :user
+      done = true
+    else
+      set_status!(
+        stats,
+        get_status(
+          nlp,
+          elapsed_time = stats.elapsed_time,
+          optimal = optimal,
+          unbounded = unbounded,
+          max_eval = max_eval,
+          iter = stats.iter,
+          max_iter = max_iter,
+          max_time = max_time,
+        ),
+      )
+      done = stats.status != :unknown
+    end
+  end
+
+  set_solution!(stats, x)
+  return stats
+end
+
+function subsolve!(R2N::R2NSolver, s, atol, n, subsolver_verbose)
+  ∇f = R2N.gx
+  cgtol = R2N.cgtol
+  H = R2N.H
+  subsolver_type = R2N.subsolver_type
+  σ = R2N.σ
+  shifted_H = R2N.shifted_H
+
+  if subsolver_type isa MinresSolver
+    minres!(
+      subsolver_type,
+      (H + σ * shifted_H), #A
```

Comment:
Is this allocating?

Reply:
yes we are trying to figure out where there is a leak